// @aries-framework/core/build/modules/routing/services/RoutingService.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RoutingService = void 0;
const AgentConfig_1 = require("../../../agent/AgentConfig");
const EventEmitter_1 = require("../../../agent/EventEmitter");
const constants_1 = require("../../../constants");
const crypto_1 = require("../../../crypto");
const plugins_1 = require("../../../plugins");
const dids_1 = require("../../dids");
const RoutingEvents_1 = require("../RoutingEvents");
const MediationRecipientService_1 = require("./MediationRecipientService");
let RoutingService = class RoutingService {
    constructor(mediationRecipientService, agentConfig, wallet, eventEmitter) {
        this.mediationRecipientService = mediationRecipientService;
        this.agentConfig = agentConfig;
        this.wallet = wallet;
        this.eventEmitter = eventEmitter;
    }
    async getRouting({ mediatorId, useDefaultMediator = true } = {}) {
        // Create and store new key
        const { verkey: publicKeyBase58 } = await this.wallet.createDid();
        const recipientKey = dids_1.Key.fromPublicKeyBase58(publicKeyBase58, crypto_1.KeyType.Ed25519);
        let routing = {
            endpoints: this.agentConfig.endpoints,
            routingKeys: [],
            recipientKey,
        };
        // Extend routing with mediator keys (if applicable)
        routing = await this.mediationRecipientService.addMediationRouting(routing, {
            mediatorId,
            useDefaultMediator,
        });
        // Emit event so other parts of the framework can react on keys created
        this.eventEmitter.emit({
            type: RoutingEvents_1.RoutingEventTypes.RoutingCreatedEvent,
            payload: {
                routing,
            },
        });
        return routing;
    }
};
RoutingService = __decorate([
    plugins_1.injectable(),
    __param(2, plugins_1.inject(constants_1.InjectionSymbols.Wallet)),
    __metadata("design:paramtypes", [MediationRecipientService_1.MediationRecipientService,
        AgentConfig_1.AgentConfig, Object, EventEmitter_1.EventEmitter])
], RoutingService);
exports.RoutingService = RoutingService;
