// @aries-framework/core/build/plugins/DependencyManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DependencyManager = void 0;
const tsyringe_1 = require("tsyringe");
class DependencyManager {
    constructor(container = tsyringe_1.container.createChildContainer()) {
        this.container = container;
    }
    registerModules(...modules) {
        modules.forEach((module) => module.register(this));
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    registerSingleton(fromOrToken, to) {
        this.container.registerSingleton(fromOrToken, to);
    }
    resolve(token) {
        return this.container.resolve(token);
    }
    registerInstance(token, instance) {
        this.container.registerInstance(token, instance);
    }
    isRegistered(token) {
        return this.container.isRegistered(token);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    registerContextScoped(token, provider) {
        if (provider)
            this.container.register(token, provider, { lifecycle: tsyringe_1.Lifecycle.ContainerScoped });
        else
            this.container.register(token, token, { lifecycle: tsyringe_1.Lifecycle.ContainerScoped });
    }
    createChild() {
        return new DependencyManager(this.container.createChildContainer());
    }
}
exports.DependencyManager = DependencyManager;
