// @aries-framework/core/build/modules/question-answer/messages/QuestionMessage.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuestionMessage = void 0;
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const AgentMessage_1 = require("../../../agent/AgentMessage");
const messageType_1 = require("../../../utils/messageType");
const models_1 = require("../models");
class QuestionMessage extends AgentMessage_1.AgentMessage {
    /**
     * Create new QuestionMessage instance.
     * @param options
     */
    constructor(options) {
        super();
        this.type = QuestionMessage.type.messageTypeUri;
        if (options) {
            this.id = options.id || this.generateId();
            this.nonce = options.nonce;
            this.questionText = options.questionText;
            this.questionDetail = options.questionDetail;
            this.signatureRequired = options.signatureRequired;
            this.validResponses = options.validResponses;
        }
    }
}
QuestionMessage.type = messageType_1.parseMessageType('https://didcomm.org/questionanswer/1.0/question');
__decorate([
    messageType_1.IsValidMessageType(QuestionMessage.type),
    __metadata("design:type", Object)
], QuestionMessage.prototype, "type", void 0);
__decorate([
    class_validator_1.IsOptional(),
    class_validator_1.IsString(),
    __metadata("design:type", String)
], QuestionMessage.prototype, "nonce", void 0);
__decorate([
    class_validator_1.IsOptional(),
    class_validator_1.IsBoolean(),
    class_transformer_1.Expose({ name: 'signature_required' }),
    __metadata("design:type", Boolean)
], QuestionMessage.prototype, "signatureRequired", void 0);
__decorate([
    class_transformer_1.Expose({ name: 'valid_responses' }),
    class_transformer_1.Type(() => models_1.ValidResponse),
    class_validator_1.ValidateNested({ each: true }),
    class_validator_1.IsInstance(models_1.ValidResponse, { each: true }),
    __metadata("design:type", Array)
], QuestionMessage.prototype, "validResponses", void 0);
__decorate([
    class_transformer_1.Expose({ name: 'question_text' }),
    class_validator_1.IsString(),
    __metadata("design:type", String)
], QuestionMessage.prototype, "questionText", void 0);
__decorate([
    class_validator_1.IsOptional(),
    class_transformer_1.Expose({ name: 'question_detail' }),
    class_validator_1.IsString(),
    __metadata("design:type", String)
], QuestionMessage.prototype, "questionDetail", void 0);
exports.QuestionMessage = QuestionMessage;
