// @aries-framework/core/build/modules/credentials/protocol/v2/handlers/V2OfferCredentialHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.V2OfferCredentialHandler = void 0;
const helpers_1 = require("../../../../../agent/helpers");
const ServiceDecorator_1 = require("../../../../../decorators/service/ServiceDecorator");
const storage_1 = require("../../../../../storage");
const V2OfferCredentialMessage_1 = require("../messages/V2OfferCredentialMessage");
class V2OfferCredentialHandler {
    constructor(credentialService, agentConfig, routingService, didCommMessageRepository) {
        this.supportedMessages = [V2OfferCredentialMessage_1.V2OfferCredentialMessage];
        this.credentialService = credentialService;
        this.agentConfig = agentConfig;
        this.routingService = routingService;
        this.didCommMessageRepository = didCommMessageRepository;
    }
    async handle(messageContext) {
        const credentialRecord = await this.credentialService.processOffer(messageContext);
        const shouldAutoRespond = await this.credentialService.shouldAutoRespondToOffer({
            credentialRecord,
            offerMessage: messageContext.message,
        });
        if (shouldAutoRespond) {
            return await this.acceptOffer(credentialRecord, messageContext);
        }
    }
    async acceptOffer(credentialRecord, messageContext, offerMessage) {
        this.agentConfig.logger.info(`Automatically sending request with autoAccept on ${this.agentConfig.autoAcceptCredentials}`);
        if (messageContext.connection) {
            const { message } = await this.credentialService.acceptOffer({
                credentialRecord,
            });
            return helpers_1.createOutboundMessage(messageContext.connection, message);
        }
        else if (offerMessage === null || offerMessage === void 0 ? void 0 : offerMessage.service) {
            const routing = await this.routingService.getRouting();
            const ourService = new ServiceDecorator_1.ServiceDecorator({
                serviceEndpoint: routing.endpoints[0],
                recipientKeys: [routing.recipientKey.publicKeyBase58],
                routingKeys: routing.routingKeys.map((key) => key.publicKeyBase58),
            });
            const recipientService = offerMessage.service;
            const { message } = await this.credentialService.acceptOffer({
                credentialRecord,
            });
            // Set and save ~service decorator to record (to remember our verkey)
            message.service = ourService;
            await this.didCommMessageRepository.saveOrUpdateAgentMessage({
                agentMessage: message,
                role: storage_1.DidCommMessageRole.Sender,
                associatedRecordId: credentialRecord.id,
            });
            return helpers_1.createOutboundServiceMessage({
                payload: message,
                service: recipientService.resolvedDidCommService,
                senderKey: ourService.resolvedDidCommService.recipientKeys[0],
            });
        }
        this.agentConfig.logger.error(`Could not automatically create credential request`);
    }
}
exports.V2OfferCredentialHandler = V2OfferCredentialHandler;
