// @aries-framework/core/build/modules/credentials/protocol/v2/handlers/V2ProposeCredentialHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.V2ProposeCredentialHandler = void 0;
const helpers_1 = require("../../../../../agent/helpers");
const V2ProposeCredentialMessage_1 = require("../messages/V2ProposeCredentialMessage");
class V2ProposeCredentialHandler {
    constructor(credentialService, agentConfig) {
        this.supportedMessages = [V2ProposeCredentialMessage_1.V2ProposeCredentialMessage];
        this.credentialService = credentialService;
        this.agentConfig = agentConfig;
    }
    async handle(messageContext) {
        const credentialRecord = await this.credentialService.processProposal(messageContext);
        const shouldAutoRespond = await this.credentialService.shouldAutoRespondToProposal({
            credentialRecord,
            proposalMessage: messageContext.message,
        });
        if (shouldAutoRespond) {
            return await this.acceptProposal(credentialRecord, messageContext);
        }
    }
    async acceptProposal(credentialRecord, messageContext) {
        this.agentConfig.logger.info(`Automatically sending offer with autoAccept on ${this.agentConfig.autoAcceptCredentials}`);
        if (!messageContext.connection) {
            this.agentConfig.logger.error('No connection on the messageContext, aborting auto accept');
            return;
        }
        const { message } = await this.credentialService.acceptProposal({ credentialRecord });
        return helpers_1.createOutboundMessage(messageContext.connection, message);
    }
}
exports.V2ProposeCredentialHandler = V2ProposeCredentialHandler;
