// @aries-framework/core/build/modules/credentials/formats/CredentialFormatService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CredentialFormatService = void 0;
const Attachment_1 = require("../../../decorators/attachment/Attachment");
const JsonEncoder_1 = require("../../../utils/JsonEncoder");
class CredentialFormatService {
    constructor(credentialRepository, eventEmitter) {
        this.credentialRepository = credentialRepository;
        this.eventEmitter = eventEmitter;
    }
    /**
     * Returns an object of type {@link Attachment} for use in credential exchange messages.
     * It looks up the correct format identifier and encodes the data as a base64 attachment.
     *
     * @param data The data to include in the attach object
     * @param id the attach id from the formats component of the message
     * @returns attachment to the credential proposal
     */
    getFormatData(data, id) {
        const attachment = new Attachment_1.Attachment({
            id,
            mimeType: 'application/json',
            data: new Attachment_1.AttachmentData({
                base64: JsonEncoder_1.JsonEncoder.toBase64(data),
            }),
        });
        return attachment;
    }
}
exports.CredentialFormatService = CredentialFormatService;
