// @aries-framework/core/build/modules/basic-messages/BasicMessagesModule.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var BasicMessagesModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasicMessagesModule = void 0;
const Dispatcher_1 = require("../../agent/Dispatcher");
const MessageSender_1 = require("../../agent/MessageSender");
const helpers_1 = require("../../agent/helpers");
const plugins_1 = require("../../plugins");
const connections_1 = require("../connections");
const handlers_1 = require("./handlers");
const repository_1 = require("./repository");
const services_1 = require("./services");
let BasicMessagesModule = BasicMessagesModule_1 = class BasicMessagesModule {
    constructor(dispatcher, basicMessageService, messageSender, connectionService) {
        this.basicMessageService = basicMessageService;
        this.messageSender = messageSender;
        this.connectionService = connectionService;
        this.registerHandlers(dispatcher);
    }
    async sendMessage(connectionId, message) {
        const connection = await this.connectionService.getById(connectionId);
        const basicMessage = await this.basicMessageService.createMessage(message, connection);
        const outboundMessage = helpers_1.createOutboundMessage(connection, basicMessage);
        await this.messageSender.sendMessage(outboundMessage);
    }
    async findAllByQuery(query) {
        return this.basicMessageService.findAllByQuery(query);
    }
    registerHandlers(dispatcher) {
        dispatcher.registerHandler(new handlers_1.BasicMessageHandler(this.basicMessageService));
    }
    /**
     * Registers the dependencies of the basic message module on the dependency manager.
     */
    static register(dependencyManager) {
        // Api
        dependencyManager.registerContextScoped(BasicMessagesModule_1);
        // Services
        dependencyManager.registerSingleton(services_1.BasicMessageService);
        // Repositories
        dependencyManager.registerSingleton(repository_1.BasicMessageRepository);
    }
};
BasicMessagesModule = BasicMessagesModule_1 = __decorate([
    plugins_1.module(),
    plugins_1.injectable(),
    __metadata("design:paramtypes", [Dispatcher_1.Dispatcher,
        services_1.BasicMessageService,
        MessageSender_1.MessageSender,
        connections_1.ConnectionService])
], BasicMessagesModule);
exports.BasicMessagesModule = BasicMessagesModule;
