// @aries-framework/core/build/modules/dids/DidsModule.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var DidsModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DidsModule = void 0;
const plugins_1 = require("../../plugins");
const repository_1 = require("./repository");
const DidResolverService_1 = require("./services/DidResolverService");
let DidsModule = DidsModule_1 = class DidsModule {
    constructor(resolverService, didRepository) {
        this.resolverService = resolverService;
        this.didRepository = didRepository;
    }
    resolve(didUrl, options) {
        return this.resolverService.resolve(didUrl, options);
    }
    resolveDidDocument(didUrl) {
        return this.resolverService.resolveDidDocument(didUrl);
    }
    findByRecipientKey(recipientKey) {
        return this.didRepository.findByRecipientKey(recipientKey);
    }
    findAllByRecipientKey(recipientKey) {
        return this.didRepository.findAllByRecipientKey(recipientKey);
    }
    /**
     * Registers the dependencies of the dids module module on the dependency manager.
     */
    static register(dependencyManager) {
        // Api
        dependencyManager.registerContextScoped(DidsModule_1);
        // Services
        dependencyManager.registerSingleton(DidResolverService_1.DidResolverService);
        dependencyManager.registerSingleton(repository_1.DidRepository);
    }
};
DidsModule = DidsModule_1 = __decorate([
    plugins_1.module(),
    plugins_1.injectable(),
    __metadata("design:paramtypes", [DidResolverService_1.DidResolverService, repository_1.DidRepository])
], DidsModule);
exports.DidsModule = DidsModule;
