// @aries-framework/core/build/modules/indy/services/IndyIssuerService.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndyIssuerService = void 0;
const AgentConfig_1 = require("../../../agent/AgentConfig");
const AriesFrameworkError_1 = require("../../../error/AriesFrameworkError");
const IndySdkError_1 = require("../../../error/IndySdkError");
const plugins_1 = require("../../../plugins");
const indyError_1 = require("../../../utils/indyError");
const IndyWallet_1 = require("../../../wallet/IndyWallet");
const IndyUtilitiesService_1 = require("./IndyUtilitiesService");
let IndyIssuerService = class IndyIssuerService {
    constructor(agentConfig, wallet, indyUtilitiesService) {
        this.indy = agentConfig.agentDependencies.indy;
        this.wallet = wallet;
        this.indyUtilitiesService = indyUtilitiesService;
        this.fileSystem = agentConfig.fileSystem;
    }
    /**
     * Create a new credential schema.
     *
     * @returns the schema.
     */
    async createSchema({ originDid, name, version, attributes }) {
        try {
            const [, schema] = await this.indy.issuerCreateSchema(originDid, name, version, attributes);
            return schema;
        }
        catch (error) {
            throw indyError_1.isIndyError(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    /**
     * Create a new credential definition and store it in the wallet.
     *
     * @returns the credential definition.
     */
    async createCredentialDefinition({ issuerDid, schema, tag = 'default', signatureType = 'CL', supportRevocation = false, }) {
        try {
            const [, credentialDefinition] = await this.indy.issuerCreateAndStoreCredentialDef(this.wallet.handle, issuerDid, schema, tag, signatureType, {
                support_revocation: supportRevocation,
            });
            return credentialDefinition;
        }
        catch (error) {
            throw indyError_1.isIndyError(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    /**
     * Create a credential offer for the given credential definition id.
     *
     * @param credentialDefinitionId The credential definition to create an offer for
     * @returns The created credential offer
     */
    async createCredentialOffer(credentialDefinitionId) {
        try {
            return await this.indy.issuerCreateCredentialOffer(this.wallet.handle, credentialDefinitionId);
        }
        catch (error) {
            throw indyError_1.isIndyError(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    /**
     * Create a credential.
     *
     * @returns Credential and revocation id
     */
    async createCredential({ credentialOffer, credentialRequest, credentialValues, revocationRegistryId, tailsFilePath, }) {
        try {
            // Indy SDK requires tailsReaderHandle. Use null if no tailsFilePath is present
            const tailsReaderHandle = tailsFilePath ? await this.indyUtilitiesService.createTailsReader(tailsFilePath) : 0;
            if (revocationRegistryId || tailsFilePath) {
                throw new AriesFrameworkError_1.AriesFrameworkError('Revocation not supported yet');
            }
            const [credential, credentialRevocationId] = await this.indy.issuerCreateCredential(this.wallet.handle, credentialOffer, credentialRequest, credentialValues, revocationRegistryId !== null && revocationRegistryId !== void 0 ? revocationRegistryId : null, tailsReaderHandle);
            return [credential, credentialRevocationId];
        }
        catch (error) {
            throw indyError_1.isIndyError(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
};
IndyIssuerService = __decorate([
    plugins_1.injectable(),
    __metadata("design:paramtypes", [AgentConfig_1.AgentConfig, IndyWallet_1.IndyWallet, IndyUtilitiesService_1.IndyUtilitiesService])
], IndyIssuerService);
exports.IndyIssuerService = IndyIssuerService;
