// @aries-framework/core/build/modules/credentials/protocol/v2/handlers/V2IssueCredentialHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.V2IssueCredentialHandler = void 0;
const helpers_1 = require("../../../../../agent/helpers");
const V2IssueCredentialMessage_1 = require("../messages/V2IssueCredentialMessage");
const V2RequestCredentialMessage_1 = require("../messages/V2RequestCredentialMessage");
class V2IssueCredentialHandler {
    constructor(credentialService, agentConfig, didCommMessageRepository) {
        this.supportedMessages = [V2IssueCredentialMessage_1.V2IssueCredentialMessage];
        this.credentialService = credentialService;
        this.agentConfig = agentConfig;
        this.didCommMessageRepository = didCommMessageRepository;
    }
    async handle(messageContext) {
        const credentialRecord = await this.credentialService.processCredential(messageContext);
        const shouldAutoRespond = await this.credentialService.shouldAutoRespondToCredential({
            credentialRecord,
            credentialMessage: messageContext.message,
        });
        if (shouldAutoRespond) {
            return await this.acceptCredential(credentialRecord, messageContext);
        }
    }
    async acceptCredential(credentialRecord, messageContext) {
        this.agentConfig.logger.info(`Automatically sending acknowledgement with autoAccept on ${this.agentConfig.autoAcceptCredentials}`);
        const requestMessage = await this.didCommMessageRepository.findAgentMessage({
            associatedRecordId: credentialRecord.id,
            messageClass: V2RequestCredentialMessage_1.V2RequestCredentialMessage,
        });
        const { message } = await this.credentialService.acceptCredential({
            credentialRecord,
        });
        if (messageContext.connection) {
            return helpers_1.createOutboundMessage(messageContext.connection, message);
        }
        else if ((requestMessage === null || requestMessage === void 0 ? void 0 : requestMessage.service) && messageContext.message.service) {
            const recipientService = messageContext.message.service;
            const ourService = requestMessage.service;
            return helpers_1.createOutboundServiceMessage({
                payload: message,
                service: recipientService.resolvedDidCommService,
                senderKey: ourService.resolvedDidCommService.recipientKeys[0],
            });
        }
        this.agentConfig.logger.error(`Could not automatically create credential ack`);
    }
}
exports.V2IssueCredentialHandler = V2IssueCredentialHandler;
