// @aries-framework/core/build/modules/dids/services/DidResolverService.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DidResolverService = void 0;
const AgentConfig_1 = require("../../../agent/AgentConfig");
const error_1 = require("../../../error");
const plugins_1 = require("../../../plugins");
const ledger_1 = require("../../ledger");
const parse_1 = require("../domain/parse");
const KeyDidResolver_1 = require("../methods/key/KeyDidResolver");
const PeerDidResolver_1 = require("../methods/peer/PeerDidResolver");
const SovDidResolver_1 = require("../methods/sov/SovDidResolver");
const WebDidResolver_1 = require("../methods/web/WebDidResolver");
const repository_1 = require("../repository");
let DidResolverService = class DidResolverService {
    constructor(agentConfig, indyLedgerService, didRepository) {
        this.logger = agentConfig.logger;
        this.resolvers = [
            new SovDidResolver_1.SovDidResolver(indyLedgerService),
            new WebDidResolver_1.WebDidResolver(),
            new KeyDidResolver_1.KeyDidResolver(),
            new PeerDidResolver_1.PeerDidResolver(didRepository),
        ];
    }
    async resolve(didUrl, options = {}) {
        this.logger.debug(`resolving didUrl ${didUrl}`);
        const result = {
            didResolutionMetadata: {},
            didDocument: null,
            didDocumentMetadata: {},
        };
        let parsed;
        try {
            parsed = parse_1.parseDid(didUrl);
        }
        catch (error) {
            return Object.assign(Object.assign({}, result), { didResolutionMetadata: { error: 'invalidDid' } });
        }
        const resolver = this.findResolver(parsed);
        if (!resolver) {
            return Object.assign(Object.assign({}, result), { didResolutionMetadata: { error: 'unsupportedDidMethod' } });
        }
        return resolver.resolve(parsed.did, parsed, options);
    }
    async resolveDidDocument(did) {
        const { didDocument, didResolutionMetadata: { error, message }, } = await this.resolve(did);
        if (!didDocument) {
            throw new error_1.AriesFrameworkError(`Unable to resolve did document for did '${did}': ${error} ${message}`);
        }
        return didDocument;
    }
    findResolver(parsed) {
        var _a;
        return (_a = this.resolvers.find((r) => r.supportedMethods.includes(parsed.method))) !== null && _a !== void 0 ? _a : null;
    }
};
DidResolverService = __decorate([
    plugins_1.injectable(),
    __metadata("design:paramtypes", [AgentConfig_1.AgentConfig, ledger_1.IndyLedgerService, repository_1.DidRepository])
], DidResolverService);
exports.DidResolverService = DidResolverService;
