// @aries-framework/core/build/storage/migration/StorageUpdateService.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var StorageUpdateService_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorageUpdateService = void 0;
const AgentConfig_1 = require("../../agent/AgentConfig");
const plugins_1 = require("../../plugins");
const StorageVersionRecord_1 = require("./repository/StorageVersionRecord");
const StorageVersionRepository_1 = require("./repository/StorageVersionRepository");
const updates_1 = require("./updates");
let StorageUpdateService = StorageUpdateService_1 = class StorageUpdateService {
    constructor(agentConfig, storageVersionRepository) {
        this.storageVersionRepository = storageVersionRepository;
        this.logger = agentConfig.logger;
    }
    async isUpToDate() {
        const currentStorageVersion = await this.getCurrentStorageVersion();
        const isUpToDate = updates_1.CURRENT_FRAMEWORK_STORAGE_VERSION === currentStorageVersion;
        return isUpToDate;
    }
    async getCurrentStorageVersion() {
        const storageVersionRecord = await this.getStorageVersionRecord();
        return storageVersionRecord.storageVersion;
    }
    async setCurrentStorageVersion(storageVersion) {
        this.logger.debug(`Setting current agent storage version to ${storageVersion}`);
        const storageVersionRecord = await this.storageVersionRepository.findById(StorageUpdateService_1.STORAGE_VERSION_RECORD_ID);
        if (!storageVersionRecord) {
            this.logger.trace('Storage upgrade record does not exist yet. Creating.');
            await this.storageVersionRepository.save(new StorageVersionRecord_1.StorageVersionRecord({
                id: StorageUpdateService_1.STORAGE_VERSION_RECORD_ID,
                storageVersion,
            }));
        }
        else {
            this.logger.trace('Storage upgrade record already exists. Updating.');
            storageVersionRecord.storageVersion = storageVersion;
            await this.storageVersionRepository.update(storageVersionRecord);
        }
    }
    /**
     * Retrieve the update record, creating it if it doesn't exist already.
     *
     * The storageVersion will be set to the INITIAL_STORAGE_VERSION if it doesn't exist yet,
     * as we can assume the wallet was created before the udpate record existed
     */
    async getStorageVersionRecord() {
        let storageVersionRecord = await this.storageVersionRepository.findById(StorageUpdateService_1.STORAGE_VERSION_RECORD_ID);
        if (!storageVersionRecord) {
            storageVersionRecord = new StorageVersionRecord_1.StorageVersionRecord({
                id: StorageUpdateService_1.STORAGE_VERSION_RECORD_ID,
                storageVersion: updates_1.INITIAL_STORAGE_VERSION,
            });
            await this.storageVersionRepository.save(storageVersionRecord);
        }
        return storageVersionRecord;
    }
};
StorageUpdateService.STORAGE_VERSION_RECORD_ID = 'STORAGE_VERSION_RECORD_ID';
StorageUpdateService = StorageUpdateService_1 = __decorate([
    plugins_1.injectable(),
    __metadata("design:paramtypes", [AgentConfig_1.AgentConfig, StorageVersionRepository_1.StorageVersionRepository])
], StorageUpdateService);
exports.StorageUpdateService = StorageUpdateService;
