// @aries-framework/core/build/modules/credentials/protocol/v1/handlers/V1ProposeCredentialHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.V1ProposeCredentialHandler = void 0;
const helpers_1 = require("../../../../../agent/helpers");
const messages_1 = require("../messages");
class V1ProposeCredentialHandler {
    constructor(credentialService, agentConfig) {
        this.supportedMessages = [messages_1.V1ProposeCredentialMessage];
        this.credentialService = credentialService;
        this.agentConfig = agentConfig;
    }
    async handle(messageContext) {
        const credentialRecord = await this.credentialService.processProposal(messageContext);
        const shouldAutoAcceptProposal = await this.credentialService.shouldAutoRespondToProposal({
            credentialRecord,
            proposalMessage: messageContext.message,
        });
        if (shouldAutoAcceptProposal) {
            return await this.acceptProposal(credentialRecord, messageContext);
        }
    }
    async acceptProposal(credentialRecord, messageContext) {
        this.agentConfig.logger.info(`Automatically sending offer with autoAccept on ${this.agentConfig.autoAcceptCredentials}`);
        if (!messageContext.connection) {
            this.agentConfig.logger.error('No connection on the messageContext, aborting auto accept');
            return;
        }
        const { message } = await this.credentialService.acceptProposal({
            credentialRecord,
        });
        return helpers_1.createOutboundMessage(messageContext.connection, message);
    }
}
exports.V1ProposeCredentialHandler = V1ProposeCredentialHandler;
