// @aries-framework/core/build/modules/credentials/protocol/v1/handlers/V1IssueCredentialHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.V1IssueCredentialHandler = void 0;
const helpers_1 = require("../../../../../agent/helpers");
const messages_1 = require("../messages");
class V1IssueCredentialHandler {
    constructor(credentialService, agentConfig, didCommMessageRepository) {
        this.supportedMessages = [messages_1.V1IssueCredentialMessage];
        this.credentialService = credentialService;
        this.agentConfig = agentConfig;
        this.didCommMessageRepository = didCommMessageRepository;
    }
    async handle(messageContext) {
        const credentialRecord = await this.credentialService.processCredential(messageContext);
        const shouldAutoRespond = await this.credentialService.shouldAutoRespondToCredential({
            credentialRecord,
            credentialMessage: messageContext.message,
        });
        if (shouldAutoRespond) {
            return await this.acceptCredential(credentialRecord, messageContext);
        }
    }
    async acceptCredential(credentialRecord, messageContext) {
        this.agentConfig.logger.info(`Automatically sending acknowledgement with autoAccept on ${this.agentConfig.autoAcceptCredentials}`);
        const { message } = await this.credentialService.acceptCredential({
            credentialRecord,
        });
        const requestMessage = await this.didCommMessageRepository.getAgentMessage({
            associatedRecordId: credentialRecord.id,
            messageClass: messages_1.V1RequestCredentialMessage,
        });
        if (messageContext.connection) {
            return helpers_1.createOutboundMessage(messageContext.connection, message);
        }
        else if (messageContext.message.service && requestMessage.service) {
            const recipientService = messageContext.message.service;
            const ourService = requestMessage.service;
            return helpers_1.createOutboundServiceMessage({
                payload: message,
                service: recipientService.resolvedDidCommService,
                senderKey: ourService.resolvedDidCommService.recipientKeys[0],
            });
        }
        this.agentConfig.logger.error(`Could not automatically create credential ack`);
    }
}
exports.V1IssueCredentialHandler = V1IssueCredentialHandler;
