// @aries-framework/core/build/modules/connections/handlers/DidExchangeResponseHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DidExchangeResponseHandler = void 0;
const helpers_1 = require("../../../agent/helpers");
const TransportDecorator_1 = require("../../../decorators/transport/TransportDecorator");
const error_1 = require("../../../error");
const OutOfBandState_1 = require("../../oob/domain/OutOfBandState");
const messages_1 = require("../messages");
const models_1 = require("../models");
class DidExchangeResponseHandler {
    constructor(agentConfig, didExchangeProtocol, outOfBandService, connectionService, didResolverService) {
        this.supportedMessages = [messages_1.DidExchangeResponseMessage];
        this.agentConfig = agentConfig;
        this.didExchangeProtocol = didExchangeProtocol;
        this.outOfBandService = outOfBandService;
        this.connectionService = connectionService;
        this.didResolverService = didResolverService;
    }
    async handle(messageContext) {
        var _a;
        const { recipientKey, senderKey, message } = messageContext;
        if (!recipientKey || !senderKey) {
            throw new error_1.AriesFrameworkError('Unable to process connection response without sender key or recipient key');
        }
        const connectionRecord = await this.connectionService.getByThreadId(message.threadId);
        if (!connectionRecord) {
            throw new error_1.AriesFrameworkError(`Connection for thread ID ${message.threadId} not found!`);
        }
        if (!connectionRecord.did) {
            throw new error_1.AriesFrameworkError(`Connection record ${connectionRecord.id} has no 'did'`);
        }
        const ourDidDocument = await this.didResolverService.resolveDidDocument(connectionRecord.did);
        if (!ourDidDocument) {
            throw new error_1.AriesFrameworkError(`Did document for did ${connectionRecord.did} was not resolved`);
        }
        // Validate if recipient key is included in recipient keys of the did document resolved by
        // connection record did
        if (!ourDidDocument.recipientKeys.find((key) => key.fingerprint === recipientKey.fingerprint)) {
            throw new error_1.AriesFrameworkError(`Recipient key ${recipientKey.fingerprint} not found in did document recipient keys.`);
        }
        const { protocol } = connectionRecord;
        if (protocol !== models_1.HandshakeProtocol.DidExchange) {
            throw new error_1.AriesFrameworkError(`Connection record protocol is ${protocol} but handler supports only ${models_1.HandshakeProtocol.DidExchange}.`);
        }
        if (!connectionRecord.outOfBandId) {
            throw new error_1.AriesFrameworkError(`Connection ${connectionRecord.id} does not have outOfBandId!`);
        }
        const outOfBandRecord = await this.outOfBandService.findById(connectionRecord.outOfBandId);
        if (!outOfBandRecord) {
            throw new error_1.AriesFrameworkError(`OutOfBand record for connection ${connectionRecord.id} with outOfBandId ${connectionRecord.outOfBandId} not found!`);
        }
        // TODO
        //
        // A connection request message is the only case when I can use the connection record found
        // only based on recipient key without checking that `theirKey` is equal to sender key.
        //
        // The question is if we should do it here in this way or rather somewhere else to keep
        // responsibility of all handlers aligned.
        //
        messageContext.connection = connectionRecord;
        const connection = await this.didExchangeProtocol.processResponse(messageContext, outOfBandRecord);
        // TODO: should we only send complete message in case of autoAcceptConnection or always?
        // In AATH we have a separate step to send the complete. So for now we'll only do it
        // if auto accept is enable
        if ((_a = connection.autoAcceptConnection) !== null && _a !== void 0 ? _a : this.agentConfig.autoAcceptConnections) {
            const message = await this.didExchangeProtocol.createComplete(connection, outOfBandRecord);
            // Disable return routing as we don't want to receive a response for this message over the same channel
            // This has led to long timeouts as not all clients actually close an http socket if there is no response message
            message.setReturnRouting(TransportDecorator_1.ReturnRouteTypes.none);
            if (!outOfBandRecord.reusable) {
                await this.outOfBandService.updateState(outOfBandRecord, OutOfBandState_1.OutOfBandState.Done);
            }
            return helpers_1.createOutboundMessage(connection, message);
        }
    }
}
exports.DidExchangeResponseHandler = DidExchangeResponseHandler;
