// @aries-framework/core/build/modules/dids/methods/peer/PeerDidResolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PeerDidResolver = void 0;
const error_1 = require("../../../../error");
const didPeer_1 = require("./didPeer");
const peerDidNumAlgo0_1 = require("./peerDidNumAlgo0");
const peerDidNumAlgo2_1 = require("./peerDidNumAlgo2");
class PeerDidResolver {
    constructor(didRepository) {
        this.supportedMethods = ['peer'];
        this.didRepository = didRepository;
    }
    async resolve(did) {
        const didDocumentMetadata = {};
        try {
            let didDocument;
            if (!didPeer_1.isValidPeerDid(did)) {
                throw new error_1.AriesFrameworkError(`did ${did} is not a valid peer did`);
            }
            const numAlgo = didPeer_1.getNumAlgoFromPeerDid(did);
            // For method 0, generate from did
            if (numAlgo === didPeer_1.PeerDidNumAlgo.InceptionKeyWithoutDoc) {
                didDocument = peerDidNumAlgo0_1.didToNumAlgo0DidDocument(did);
            }
            // For Method 1, retrieve from storage
            else if (numAlgo === didPeer_1.PeerDidNumAlgo.GenesisDoc) {
                const didDocumentRecord = await this.didRepository.getById(did);
                if (!didDocumentRecord.didDocument) {
                    throw new error_1.AriesFrameworkError(`Found did record for method 1 peer did (${did}), but no did document.`);
                }
                didDocument = didDocumentRecord.didDocument;
            }
            // For Method 2, generate from did
            else {
                didDocument = peerDidNumAlgo2_1.didToNumAlgo2DidDocument(did);
            }
            return {
                didDocument,
                didDocumentMetadata,
                didResolutionMetadata: { contentType: 'application/did+ld+json' },
            };
        }
        catch (error) {
            return {
                didDocument: null,
                didDocumentMetadata,
                didResolutionMetadata: {
                    error: 'notFound',
                    message: `resolver_error: Unable to resolve did '${did}': ${error}`,
                },
            };
        }
    }
}
exports.PeerDidResolver = PeerDidResolver;
